sw/device/lib/dif/dif_clkmgr.c

```diff
@@ -3,7 +3,131 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "sw/device/lib/dif/dif_clkmgr.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
 
 #include "clkmgr_regs.h"  // Generated
 
-#error "Clock Manager DIF Implementation Incomplete"
+/**
+ * There are fewer Gateable clocks than this in the hardware, but this is the
+ * threshold at which we will need to change the implementation of the DIF.
+ */
+static const uint32_t kMaxGateableClocks = CLKMGR_PARAM_REG_WIDTH;
+
+/**
+ * There are fewer Hintable clocks than this in the hardware, but this is the
+ * threshold at which we will need to change the implementation of the DIF.
+ */
+static const uint32_t kMaxHintableClocks = CLKMGR_PARAM_REG_WIDTH;
+
+dif_clkmgr_result_t dif_clkmgr_init(dif_clkmgr_params_t params,
+                                    dif_clkmgr_t *handle) {
+  if (handle == NULL) {
+    return kDifClkmgrBadArg;
+  }
+
+  handle->params = params;
+  return kDifClkmgrOk;
+}
+
+dif_clkmgr_result_t dif_clkmgr_gateable_clock_get_enabled(
+    const dif_clkmgr_t *handle, dif_clkmgr_gateable_clock_t clock,
+    bool *is_enabled) {
+  if (handle == NULL || is_enabled == NULL || clock >= kMaxGateableClocks) {
+    return kDifClkmgrBadArg;
+  }
+
+  uint32_t clk_enables_val = mmio_region_read32(handle->params.base_addr,
+                                                CLKMGR_CLK_ENABLES_REG_OFFSET);
+  *is_enabled = bitfield_bit32_read(clk_enables_val, clock);
+
+  return kDifClkmgrOk;
+}
+
+dif_clkmgr_result_t dif_clkmgr_gateable_clock_set_enabled(
+    const dif_clkmgr_t *handle, dif_clkmgr_gateable_clock_t clock,
+    dif_clkmgr_toggle_t new_state) {
+  if (handle == NULL || clock >= kMaxGateableClocks) {
+    return kDifClkmgrBadArg;
+  }
+
+  bool new_clk_enables_bit;
+  switch (new_state) {
+    case kDifClkmgrToggleEnabled:
+      new_clk_enables_bit = true;
+      break;
+    case kDifClkmgrToggleDisabled:
+      new_clk_enables_bit = false;
+      break;
+    default:
+      return kDifClkmgrBadArg;
```

Comment:
Not sure whether `kDifClkmgrBadArg` or `kDifClkmgrError` is better suited here. I think `kDifClkmgrBadArg` is probably the right error code as it would cover both variants (even if the value is glitched, it still is a bad arg).

Reply:
BadArg is allowed here, as we haven't yet accessed an MMIO - I could see a different implementation requiring Error here.